src/cached_object_store.rs

```diff
@@ -762,6 +797,224 @@ impl LocalCacheEntry for FsCacheEntry {
+        }
+
+        // if the cache size exceeds the limit, evict the cache files in batch with the batch_factor,
+        // this may help to avoid the cases like triggering the evictor too frequently when the cache
+        // size is just slightly above the limit.
+        let mut total_bytes: u64 = 0;
+        for _ in 0..self.batch_factor {
+            let evicted_bytes = self.evict_once().await;
+            if evicted_bytes == 0 {
+                return total_bytes;
+            }
+            self.tracked_bytes
+                .fetch_sub(evicted_bytes, Ordering::SeqCst);
+            total_bytes += evicted_bytes;
+        }
+
+        total_bytes
+    }
+
+    // evict once, return the bytes evicted
+    async fn evict_once(&mut self) -> u64 {
+        let (target, target_bytes) = match self.pick_evict_target().await {
+            Some(target) => target,
+            None => return 0,
+        };
+
+        match tokio::fs::remove_file(&target).await {
+            Err(err) => {
+                warn!("evictor: failed to remove the cache file: {}", err);
+                0
+            }
+            Ok(()) => {
+                debug!(
+                    "evictor: evicted cache file: {:?}, bytes: {}",
+                    target, target_bytes
+                );
+                target_bytes
+            }
+        }
+    }
+
+    pub async fn scan_cache_folder_bytes(
+        root_folder: &std::path::PathBuf,
+    ) -> Result<usize, Box<dyn std::error::Error>> {
+        let mut total_bytes = 0;
+        let iter = WalkDir::new(root_folder).into_iter();
+        for entry in iter {
+            let entry = entry?;
+            let metadata = tokio::fs::metadata(entry.path()).await?;
+            total_bytes += metadata.len();
+        }
+        Ok(total_bytes as usize)
+    }
+
+    // pick a file to evict, return None if no file is picked. it takes a pick-of-2 strategy, randomized pick
+    // two of files, compare their last access time, and evict the older one.
+    async fn pick_evict_target(&mut self) -> Option<(std::path::PathBuf, u64)> {
+        // extend the evict_buffer if it's empty. it will randomized iterate the cache folder,
+        // and randomly pick 1000 files into the evict_buffer.
+        let mut rng = rand::rngs::StdRng::from_entropy();
+        if self.evict_buffer.len() < 2 {
+            let iter = WalkDir::new(&self.root_folder).into_iter();
+            for entry in iter.choose_multiple(&mut rng, 1000) {
+                let entry = match entry {
+                    Ok(entry) => entry,
+                    Err(err) => {
+                        warn!("evictor: unexpected error on reading dir entry to pick the evict target, skip the entry: {}", err);
+                        break;
+                    }
+                };
+
+                if entry.file_type().is_dir() {
+                    continue;
+                }
+                self.evict_buffer.push_back(entry.path().to_path_buf());
```

Comment:
do we need to make sure that the entries are not already present in the evict buffer, which is possible when the evict buffer is down to 1 entry? Otherwise we'll eventually double-evict that entry, which isn't a huge deal but better to clean it up if I'm reading this right.

Reply:
evict_buffer is to store the random sampled files in a batch, yes, it's possible to sampled the same file twice times in the buffer, but given the probability, the side effect will be limitted.

we can also consider changing it into a HashSet, and random pick two from them on every eviction is executed.

Reply:
Yeah that makes sense. I'm ok with leaving it as is - seems simpler.